### ISSUE TITLE
Wekan, Mediagoblin installations fail with "Package hash did not match"

### ISSUE THREAD

**Issue by rlafuente** [sandstorm-io/sandstorm#1388]
All parts of the Sandstorm installation went fine, and it's actually _fun_ to set up. Thanks everyone for making the process so smooth!

I did run into a problem on my self-hosted instance: certain apps throw an error when trying to install -- `Package hash did not match`. This happens with Wekan and Mediagoblin. All Ether* apps, Draw.io, Gitweb and a few others I tried install fine.

Relevant portion from `sandstorm.log`:

```
Downloading app: https://app-index.sandstorm.io/packages/7c76eaa536d4a28742e965ebccc1dc89
Failed to install app: Error: Package hash did not match.
    at AppInstaller.<anonymous> (server/installer.js:492:1)
    at IncomingMessage.<anonymous> (server/installer.js:365:1)
    at IncomingMessage.emit (events.js:117:20)
    at _stream_readable.js:944:16
    at process._tickCallback (node.js:448:13)
Downloading app: https://app-index.sandstorm.io/packages/35f82e4bf3ef0683558336494a3dc9d6
App ready: 35f82e4bf3ef0683558336494a3dc9d6
fetching keybase 5FDCAA736BE1402EDA0107A4F157DCE30ECD34E4
fetching keybase 5FDCAA736BE1402EDA0107A4F157DCE30ECD34E4
Downloading app: https://app-index.sandstorm.io/packages/54240bdb86c60e44b2bf0437561e236f
Failed to install app: Error: Package hash did not match.
    at AppInstaller.<anonymous> (server/installer.js:492:1)
    at IncomingMessage.<anonymous> (server/installer.js:365:1)
    at IncomingMessage.emit (events.js:117:20)
    at _stream_readable.js:944:16
    at process._tickCallback (node.js:448:13)
```

**Comment by paulproteus** [comment]
Interesting! You're not the only person who's reported an issue like this.
I'd love to get to the bottom of it.

Please run:

wget
https://app-index.sandstorm.io/packages/7c76eaa536d4a28742e965ebccc1dc89
spk verify 7c76eaa536d4a28742e965ebccc1dc89


and see if that seems to print some JSON as output. If it does, then I
wonder what's going on!

**Comment by rlafuente** [comment]
Okay, unexpected result that might yield a clue: the wget process gets repeatedly terminated between 20-30% wth the error: `Read error at byte 40255184/119446716 (The TLS connection was non-properly terminated.). Retrying.`

This reminds me of a little detail that I thought irrelevant when posting the bug -- when trying to install from the Sandstorm App repository, the progress bar would hit around 1/3, then the "hash did not match" error would pop up. I'd venture to say that the download ends up incomplete (maybe a server misconfig?) and obviously doesn't match the package hash.

**Comment by paulproteus** [comment]
Fascinating. So then the question is - who is monkeying with this HTTPS
connection? It works for me.

I wonder if this relates to our use of CloudFlare.

You can test that by adding this line to your /etc/hosts

107.178.223.140 app-index.sandstorm.io



then re-doing the wget

**Comment by kentonv** [comment]
The error indicates the TCP connection is being dropped prematurely (no one is "monkeying with HTTPS"). Probably there's a bad router or physical link somewhere.

We should probably fix Sandstorm to reconnect and issue a range request (and we should implement range requests in the static web publishing path).

**Comment by paulproteus** [comment]
Adding usability label since it prevents people from using Sandstorm in a way that provides no information on how users can help themselves. (Admittedly they can't, so it's a deeper issue that has usability consequences.)

**Comment by rlafuente** [comment]
Okay, so after adding the line you indicated to `/etc/hosts` and retrying wget, the download terminates at around 2% instead of ~25% with a different error:

    Read error at byte 2260688/119446716 (A TLS packet with unexpected length was received.). Retrying.

I tried this on two different machines in distinct networks, and had the same behaviour.

Thanks for such a quick response, guys! Hope I can help to get to the bottom of this, really enjoying my Sandstorm experience.

**Comment by paulproteus** [comment]
Out of curiosity, what is your Internet connectivity?​ It strikes me as
strange that the download keeps getting terminated.

**Comment by rlafuente** [comment]
I'm pretty stumped after trying out some wget options and always the same result. My connectivity doesn't seem to be the issue (tested on 100Mbps and 25Mbps connections), but I wonder if it's a router or ISP problem on my side? The thing is that the progress is somewhat consistent, always breaking between 20 and 30 percent of the download.

I'm running Debian Jessie, forgot to mention that detail.

**Comment by paulproteus** [comment]
I presume it's some ISP problem.​ It's pretty tragic, but you're definitely
not the only person who's seen this problem, so it does seem to me
something that Sandstorm should work around.

You might consider asking your ISP what's up. If you want to get fancy, you
can do some packet sniffing to see if they're doing TCP RST spoofing; see
e.g. https://www.eff.org/wp/packet-forgery-isps-report-comcast-affair

**Comment by jempson** [comment]
I work for my ISP and have been hosting web sites on my modem for a couple of years now.  This is the first time I've ever received a disconnect when attempting to connect to a web server.

I installed sandstorm about 20 minutes ago and have since been unable to install a single app.  I followed your instructions to attempt the download and verification.  I get disconnected at 0% whether I have the /etch/hosts entry or not.

--2016-01-22 16:11:45--  (try:20)  https://app-index.sandstorm.io/packages/7c76eaa536d4a28742e965ebccc1dc89
Connecting to app-index.sandstorm.io (app-index.sandstorm.io)|107.178.223.140|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 119446716 (114M) [application/octet-stream]
Saving to: ‘7c76eaa536d4a28742e965ebccc1dc89.1’

 0% [                                                                                                                                     ] 65,232      --.-K/s   in 0.07s   

2016-01-22 16:11:45 (924 KB/s) - Connection closed at byte 65232. Giving up.

traceroute app-index.sandstorm.io
traceroute to app-index.sandstorm.io (104.25.219.6), 30 hops max, 60 byte packets
 1  10.67.0.1 (10.67.0.1)  15.417 ms  15.605 ms  15.633 ms
 2  dynamic-74.81.100.61.zitomedia.net (74.81.100.61)  22.998 ms  23.936 ms  23.970 ms
 3  tr0-atlnga00.zitomedia.net (74.81.98.236)  44.037 ms  45.123 ms  44.982 ms
 4  198.32.132.136 (198.32.132.136)  51.143 ms  49.104 ms  51.264 ms
 5  104.25.219.6 (104.25.219.6)  50.079 ms  47.988 ms  50.178 ms

And as you can see I have no problem mirroring your main site with wget -mr.

wget -mr https://sandstorm.io/
--2016-01-22 16:17:19--  https://sandstorm.io/
Resolving sandstorm.io (sandstorm.io)... 104.25.219.6, 104.25.218.6
Connecting to sandstorm.io (sandstorm.io)|104.25.219.6|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [text/html]
Saving to: ‘sandstorm.io/index.html’

    [ <=>                                                                                                                                 ] 9,935       --.-K/s   in 0s      

2016-01-22 16:17:19 (70.1 MB/s) - ‘sandstorm.io/index.html’ saved [9935]

Loading robots.txt; please ignore errors.
--2016-01-22 16:17:19--  https://sandstorm.io/robots.txt
Reusing existing connection to sandstorm.io:443.
HTTP request sent, awaiting response... 404 Not Found
2016-01-22 16:17:19 ERROR 404: Not Found.

--2016-01-22 16:17:19--  https://sandstorm.io/style.css
Reusing existing connection to sandstorm.io:443.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [text/css]
Saving to: ‘sandstorm.io/style.css’

    [ <=>                                                                                                                                 ] 55,918      --.-K/s   in 0.06s   

2016-01-22 16:17:19 (913 KB/s) - ‘sandstorm.io/style.css’ saved [55918]

--2016-01-22 16:17:19--  https://sandstorm.io/feed.xml
Reusing existing connection to sandstorm.io:443.
HTTP request sent, awaiting response... 200 OK
Length: 44525 (43K) [text/xml]
Saving to: ‘sandstorm.io/feed.xml’

100%[====================================================================================================================================>] 44,525      --.-K/s   in 0.003s  

2016-01-22 16:17:19 (12.7 MB/s) - ‘sandstorm.io/feed.xml’ saved [44525/44525]

--2016-01-22 16:17:19--  https://sandstorm.io/get
Reusing existing connection to sandstorm.io:443.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [text/html]
Saving to: ‘sandstorm.io/get’

    [ <=>                                                                                                                                 ] 7,025       --.-K/s   in 0s      

2016-01-22 16:17:19 (33.5 MB/s) - ‘sandstorm.io/get’ saved [7025]

--2016-01-22 16:17:19--  https://sandstorm.io/features
Reusing existing connection to sandstorm.io:443.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [text/html]
Saving to: ‘sandstorm.io/features’

    [ <=>                                                                                                                                 ] 10,251      --.-K/s   in 0.001s  

2016-01-22 16:17:19 (9.21 MB/s) - ‘sandstorm.io/features’ saved [10251]


Basically you can see the only issue I have is connecting to the app-index server.

**Comment by paulproteus** [comment]
Yeah, ​it turns out this looks like a problem with our server! We're
looking into it now.

**Comment by paulproteus** [comment]
Huge thanks for alerting us to the problem.​ Sorry on our part for
believing this was a network issue on your end (although it did work for
me...!).

It seems that the connection termination might have related to which cipher
suite was getting used, which is why it wasn't as easy to reproduce,
because it related to our nginx configuration being unable to write some
files related to TLS state.

**Comment by jempson** [comment]
That was awesomely fast response time.  Apps are now installing.

Thank you.

**Comment by paulproteus** [comment]
@jempson you are very patient and you are very welcome!

@kentonv can reply with more information, if he wants, since he is the one who actually fixed the server.

**Comment by kentonv** [comment]
It was the old "ran out of disk space, now things fail in mysterious ways". I wasn't expecting to run out of disk space on the nginx machine so didn't have it properly monitored. :(

I doubt this solved @rlafuente's problem, though, as the disk space problem just came up today.

**Comment by rlafuente** [comment]
Hey, here's some good news then -- it's all working as expected now! The `wget` call runs without a hitch (without the `/etc/hosts` modification) and the packages I mentioned in the initial report now install properly. 

Thanks, guys!

**Comment by paulproteus** [comment]
I'll close this as @rlafuente reports that thing are OK.

**Comment by paulproteus** [comment]
Thanks for reporting this issue!
